src/Xamarin.Android.Build.Tasks/Tasks/CreateAar.cs

```diff
@@ -0,0 +1,106 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Build.Framework;
+using Xamarin.Tools.Zip;
+
+namespace Xamarin.Android.Tasks
+{
+	public class CreateAar : AndroidTask
+	{
+		public override string TaskPrefix => "CAAR";
+
+		public ITaskItem [] AndroidAssets { get; set; }
+
+		public ITaskItem [] AndroidResources { get; set; }
+
+		public ITaskItem [] AndroidEnvironment { get; set; }
+
+		public ITaskItem [] JarFiles { get; set; }
+
+		public ITaskItem [] NativeLibraries { get; set; }
+
+		[Required]
+		public string AssetDirectory { get; set; }
+
+		[Required]
+		public string OutputFile { get; set; }
+
+		public override bool RunTask ()
+		{
+			Directory.CreateDirectory (Path.GetDirectoryName (OutputFile));
+
+			using (var stream = File.Create (OutputFile))
+			using (var aar = ZipArchive.Open (stream)) {
+				var existingEntries = new HashSet<string> (StringComparer.Ordinal);
+				foreach (var entry in aar) {
+					Log.LogDebugMessage ("Existing entry: " + entry.FullName);
+					existingEntries.Add (entry.FullName);
+				}
+				if (AndroidAssets != null) {
+					foreach (var asset in AndroidAssets) {
+						// See: https://github.com/xamarin/xamarin-android/commit/665cb59205f8ac565b6acbda740624844bc1cbd9
+						if (Directory.Exists (asset.ItemSpec)) {
+							Log.LogDebugMessage ($"Skipping item, is a directory: {asset.ItemSpec}");
+							continue;
+						}
+						var relative = MonoAndroidHelper.GetRelativePathForAndroidAsset (AssetDirectory, asset);
+						var archivePath = "assets/" + relative.Replace ('\\', '/');
+						aar.AddStream (File.OpenRead (asset.ItemSpec), archivePath);
+						existingEntries.Remove (archivePath);
+					}
+				}
+				if (AndroidResources != null) {
+					foreach (var resource in AndroidResources) {
+						// See: https://github.com/xamarin/xamarin-android/commit/665cb59205f8ac565b6acbda740624844bc1cbd9
+						if (Directory.Exists (resource.ItemSpec)) {
+							Log.LogDebugMessage ($"Skipping item, is a directory: {resource.ItemSpec}");
+							continue;
+						}
+						var directory = Path.GetDirectoryName (resource.ItemSpec);
+						var archivePath = "res/" + Path.GetFileName (directory) + "/" + Path.GetFileName (resource.ItemSpec);
+						aar.AddStream (File.OpenRead (resource.ItemSpec), archivePath);
+						existingEntries.Remove (archivePath);
+					}
+				}
+				if (AndroidEnvironment != null) {
+					foreach (var env in AndroidEnvironment) {
+						var archivePath = ".netenv/" + Path.GetFileName (env.ItemSpec);
```

Comment:
While this is presumably not "standard practice", nothing prevents someone from using `@(AndroidEnvironment)` on a file in a directory:

```xml
<AndroidEnvironment Include="Env.txt" />
<AndroidEnvironment Include="Dir1/Env.txt" />
<AndroidEnvironment Include="Dir2/Env.txt" />
```

We thus probably need to "hash" the filenames, much as we do assembly names & namespaces to compute JCW package names, so that we don't have inadvertent collisions.  Or throw the directory name in there, somehow.

Reply:
I changed this to:
```csharp
var archivePath = $".netenv/{Files.HashString (env.ItemSpec)}.env";
```

Reply:
I also had to replace `\` with `/`, so we should get the same hash values on all platforms.